### DISCUSSION TITLE
Copilot generates insecure code by default

### DISCUSSION THREAD

**Discussion by bannsec** [community/community#6636] [Copilot Conversations] [answered]
So a lot of what copilot does is amazing. However, it seems to not care about any security related concerns. For instance, there is no accountability for sizes being read. What follows is fully co-pilot generated. I only created the comments and let copilot create the code.

```C
#include <stdio.h>
#include <stdlib.h>

// Structure:
// Name: user
// Description: This is a structure that holds the user's information.
// Fields:
// Name: name
// Description: This is the name of the user.
// Type: string
// Name: age
// Description: This is the age of the user.
// Type: int
// Name: address
// Description: This is the address of the user.
// Type: string
// Name: Favorite_color
// Description: This is the favorite color of the user.
// Type: string
struct user
{
    char name[20];
    int age;
    char address[20];
    char Favorite_color[20];
};

// Function: get_user_info
// Prompts the user for their information and stores it in the structure.
// Returns: Pointer to the structure
struct user *get_user_info(void)
{
    struct user *user_info;
    user_info = (user *) malloc(sizeof(struct user));
    printf("Enter your name: ");
    scanf("%s", user_info->name);
    printf("Enter your age: ");
    scanf("%d", &user_info->age);
    printf("Enter your address: ");
    scanf("%s", user_info->address);
    printf("Enter your favorite color: ");
    scanf("%s", user_info->Favorite_color);
    return user_info;
}

// Function: main
// Description: This is the main function.
// Gets user's information and prints it out.
int main(void)
{
    struct user *user_info;
    user_info = get_user_info();
    printf("Name: %s\n", user_info->name);
    printf("Age: %d\n", user_info->age);
    printf("Address: %s\n", user_info->address);
    printf("Favorite color: %s\n", user_info->Favorite_color);
    return 0;
}
```

It functionally works. However, note that it decided in the struct to create a size of 20 for the buffers, but when it reads in the struct it does absolutely no bounds check. This means by default, users will be creating insecure code. This could quickly get out of hand if copilot really catches on for coding assistance, so it would be in everyones best interest to somehow train copilot to check buffer sizes prior to just reading in unlimited data.

**Comment by hsenag** [comment, marked as answer]
Hi, thanks a lot for the feedback. We have a [FAQ entry about insecure code](https://copilot.github.com/#faq-can-github-copilot-introduce-insecure-code-in-its-suggestions) and as we say there, as we improve GitHub Copilot we will work to reduce this problem.

**Comment by Kremnari** [comment]
Would it be possible to have copilot provide a security rating for generated content, at least in the interim?  A way of showing that copilot at least recognizes what code needs a security review?

**Reply by Urethramancer** [reply]
It's trained on what most people write, i.e. awful, awful, dreadful code ;)

A "rate this result" feature could help improve the quality, but perhaps security scans before picking code to train on is better.
